File: android/app/src/main/java/com/herobrawl/game/data/Factions.kt
package com.herobrawl.game.data

import com.herobrawl.game.model.Faction
import com.herobrawl.game.model.FactionId

object Factions {
    val all: Map<FactionId, Faction> = mapOf(
        FactionId.VANGUARD to Faction(
            FactionId.VANGUARD, "Vanguard", 0xFF4AA3FF, 0xFF82C4FF,
            "Disciplined knights of the crystal citadel.",
            listOf(FactionId.HORDE)
        ),
        FactionId.HORDE to Faction(
            FactionId.HORDE, "Horde", 0xFFFF8A3D, 0xFFFFB370,
            "Tribal warriors from the scorched badlands.",
            listOf(FactionId.WILDWOOD)
        ),
        FactionId.WILDWOOD to Faction(
            FactionId.WILDWOOD, "Wildwood", 0xFF8AD65F, 0xFFB6EA8F,
            "Druids and beastkin of the old forests.",
            listOf(FactionId.ARCANE)
        ),
        FactionId.ARCANE to Faction(
            FactionId.ARCANE, "Arcane", 0xFFB47DFF, 0xFFD3B1FF,
            "Rune-etched sorcerers and starborn scholars.",
            listOf(FactionId.VANGUARD)
        ),
        FactionId.RADIANCE to Faction(
            FactionId.RADIANCE, "Radiance", 0xFFFFD966, 0xFFFFE89E,
            "Seraphic champions of the dawn.",
            listOf(FactionId.ABYSS)
        ),
        FactionId.ABYSS to Faction(
            FactionId.ABYSS, "Abyss", 0xFFE44A7A, 0xFFFF7AA3,
            "Infernal lords and shadow-sworn sovereigns.",
            listOf(FactionId.RADIANCE)
        ),
    )

    data class Aura(
        val label: String,
        val attack: Double,
        val health: Double,
        val speed: Double,
    )

    fun lineupAura(factions: List<FactionId>): Aura {
        if (factions.isEmpty()) return Aura("Empty lineup", 0.0, 0.0, 0.0)
        val counts = factions.groupingBy { it }.eachCount()
        val max = counts.maxByOrNull { it.value }
        val rad = counts[FactionId.RADIANCE] ?: 0
        val aby = counts[FactionId.ABYSS] ?: 0
        if (rad >= 2 && aby >= 2 && factions.size == 5) {
            return Aura("Dawnfall Pact · 2 Radiance + 2 Abyss", 0.18, 0.20, 0.08)
        }
        if (max != null && max.value == 5) {
            return when (max.key) {
                FactionId.VANGUARD -> Aura("Mono-Vanguard Aura", 0.10, 0.30, 0.00)
                FactionId.HORDE -> Aura("Mono-Horde Aura", 0.22, 0.10, 0.05)
                FactionId.WILDWOOD -> Aura("Mono-Wildwood Aura", 0.10, 0.20, 0.12)
                FactionId.ARCANE -> Aura("Mono-Arcane Aura", 0.25, 0.05, 0.05)
                FactionId.RADIANCE -> Aura("Mono-Radiance Aura", 0.12, 0.20, 0.08)
                FactionId.ABYSS -> Aura("Mono-Abyss Aura", 0.20, 0.15, 0.08)
            }
        }
        if (max != null && max.value == 4) {
            val name = all.getValue(max.key).name
            return Aura("$name Bond (4)", 0.08, 0.10, 0.03)
        }
        return Aura("Balanced lineup", 0.0, 0.0, 0.0)
    }

    fun advantage(attacker: FactionId, defender: FactionId): Double = when {
        defender in all.getValue(attacker).strongVs -> 0.20
        attacker in all.getValue(defender).strongVs -> -0.15
        else -> 0.0
    }
}

File: src/game/factions.ts
import type { Faction, FactionId } from "../types";

// Rock-paper-scissors cycle + mirror pair, similar in spirit to TapTap Heroes
// but rebalanced for HeroBrawl with transparent advantage values.
export const FACTIONS: Record<FactionId, Faction> = {
  vanguard: {
    id: "vanguard",
    name: "Vanguard",
    color: "#4aa3ff",
    accent: "#82c4ff",
    lore: "Disciplined knights of the crystal citadel.",
    strongVs: ["horde"],
  },
  horde: {
    id: "horde",
    name: "Horde",
    color: "#ff8a3d",
    accent: "#ffb370",
    lore: "Tribal warriors from the scorched badlands.",
    strongVs: ["wildwood"],
  },
  wildwood: {
    id: "wildwood",
    name: "Wildwood",
    color: "#8ad65f",
    accent: "#b6ea8f",
    lore: "Druids and beastkin of the old forests.",
    strongVs: ["arcane"],
  },
  arcane: {
    id: "arcane",
    name: "Arcane",
    color: "#b47dff",
    accent: "#d3b1ff",
    lore: "Rune-etched sorcerers and starborn scholars.",
    strongVs: ["vanguard"],
  },
  radiance: {
    id: "radiance",
    name: "Radiance",
    color: "#ffd966",
    accent: "#ffe89e",
    lore: "Seraphic champions of the dawn.",
    strongVs: ["abyss"],
  },
  abyss: {
    id: "abyss",
    name: "Abyss",
    color: "#e44a7a",
    accent: "#ff7aa3",
    lore: "Infernal lords and shadow-sworn sovereigns.",
    strongVs: ["radiance"],
  },
};

export const FACTION_IDS: FactionId[] = [
  "vanguard",
  "horde",
  "wildwood",
  "arcane",
  "radiance",
  "abyss",
];

// Aura bonus: 5 of the same faction in lineup grants a strong bonus.
export function lineupAura(factions: FactionId[]): {
  faction: FactionId | "mixed" | "dawnfall";
  description: string;
  bonus: { attack: number; health: number; speed: number };
} {
  const counts = new Map<FactionId, number>();
  for (const f of factions) counts.set(f, (counts.get(f) ?? 0) + 1);

  const maxEntry = [...counts.entries()].sort((a, b) => b[1] - a[1])[0];

  // Dawnfall pact: 2 Radiance + 2 Abyss + any 1 = big balanced bonus
  const rad = counts.get("radiance") ?? 0;
  const aby = counts.get("abyss") ?? 0;
  if (rad >= 2 && aby >= 2 && factions.length === 5) {
    return {
      faction: "dawnfall",
      description: "Dawnfall Pact · 2 Radiance + 2 Abyss",
      bonus: { attack: 0.18, health: 0.2, speed: 0.08 },
    };
  }

  if (maxEntry && maxEntry[1] === 5) {
    const bonuses: Record<FactionId, { attack: number; health: number; speed: number }> = {
      vanguard: { attack: 0.1, health: 0.3, speed: 0 },
      horde: { attack: 0.22, health: 0.1, speed: 0.05 },
      wildwood: { attack: 0.1, health: 0.2, speed: 0.12 },
      arcane: { attack: 0.25, health: 0.05, speed: 0.05 },
      radiance: { attack: 0.12, health: 0.2, speed: 0.08 },
      abyss: { attack: 0.2, health: 0.15, speed: 0.08 },
    };
    return {
      faction: maxEntry[0],
      description: `Mono-${FACTIONS[maxEntry[0]].name} Aura`,
      bonus: bonuses[maxEntry[0]],
    };
  }

  if (maxEntry && maxEntry[1] === 4) {
    return {
      faction: maxEntry[0],
      description: `${FACTIONS[maxEntry[0]].name} Bond (4)`,
      bonus: { attack: 0.08, health: 0.1, speed: 0.03 },
    };
  }

  return {
    faction: "mixed",
    description: "Balanced lineup",
    bonus: { attack: 0, health: 0, speed: 0 },
  };
}

// +20% dmg, +10% speed when attacker's faction is strong vs defender's
export function factionAdvantage(
  attacker: FactionId,
  defender: FactionId
): { damage: number; speed: number } {
  if (FACTIONS[attacker].strongVs.includes(defender)) {
    return { damage: 0.2, speed: 0.1 };
  }
  if (FACTIONS[defender].strongVs.includes(attacker)) {
    return { damage: -0.15, speed: -0.05 };
  }
  return { damage: 0, speed: 0 };
}
